## PACKAGES

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 1. IMPORTING DATA FROM SQL SERVER

### connecting to the database

In [5]:
from sqlalchemy import create_engine
from sqlalchemy.engine import URL

In [6]:
import psycopg2
hostname = 'localhost'
database = 'nyfd'
user_name = 'postgres'
pwd = 'raksha' 
port_id= 5432
conn=None
cur=None

In [7]:
engine = create_engine(f'postgresql://{user_name}:{pwd}@{hostname}:{port_id}/{database}')

In [8]:
sql = "SELECT * FROM public.insp"

In [9]:
insp = pd.read_sql_query(sql,engine)
insp.sample(10)

facility  \
504047                      BOWL-O-DROME   
1227441  DUNKIN' (WEBSTER-HOLT RD MOBIL)   
297964                    PICCOLO'S CAFE   
344293                       HUA YU INC.   
284060        TACO PROJECT PLEASANTVILLE   
344333                         SILO, THE   
1148111                    UNITED BUFFET   
270711         I GO INN RESTAURANT & BAR   
489026                      BEACHCOMBERS   
57184               RAVAGH PERSIAN GRILL   

                                      address date_of_inspection  \
504047       2712 PINE AVENUE,  NIAGARA FALLS           3/6/2013   
1227441               873 HOLT ROAD,  WEBSTER          3/10/2010   
297964           161 Main STREET,  BINGHAMTON          3/14/2024   
344293             2766 MAIN STREET,  NEWFANE           4/4/2016   
284060       465 BEDFORD ROAD,  PLEASANTVILLE          5/10/2024   
344333         537 AVIATION ROAD,  QUEENSBURY          1/29/2015   
1148111       2400  WATT STREET,  SCHENECTADY           4/9/2015   
270711     241 SOUTH SHORE  ROAD,  NORTHVILLE           8/1/2011   
489026          5909 WEST LAKE ROAD,  CONESUS           7/1/2022   
57184    210 MINEOLA  AVENUE,  ROSLYN HEIGHTS           4/5/2010   

        violation_item                              violation_description  \
504047              2E  Accurate thermometers not available or used to...   
1227441            15B  Lighting and ventilation inadequate, fixtures ...   
297964             15B  Lighting and ventilation inadequate, fixtures ...   
344293              5C  Potentially hazardous foods are not stored und...   
284060              8A  Food not protected during storage, preparation...   
344333              5C  Potentially hazardous foods are not stored und...   
1148111             5A  Potentially hazardous foods are not kept at or...   
270711              3C  Food workers do not use proper utensils to eli...   
489026            None                                               None   
57184              12C  Plumbing and sinks not properly sized, install...   

             critical_violation  total_critical_violations  \
504047       Critical Violation                        1.0   
1227441  Not Critical Violation                        0.0   
297964   Not Critical Violation                        1.0   
344293       Critical Violation                        1.0   
284060   Not Critical Violation                        0.0   
344333       Critical Violation                        9.0   
1148111      Critical Violation                        3.0   
270711       Critical Violation                       12.0   
489026   Not Critical Violation                        0.0   
57184    Not Critical Violation                        0.0   

         total_crit_not_corrected  total_noncritical_violations  \
504047                        0.0                           3.0   
1227441                       0.0                           3.0   
297964                        0.0                           9.0   
344293                        0.0                           1.0   
284060                        0.0                           6.0   
344333                        0.0                          28.0   
1148111                       0.0                           8.0   
270711                        0.0                          25.0   
489026                        0.0                           0.0   
57184                         0.0                           7.0   

             local_health_department  ...         permitted_dba  \
504047                Niagara County  ...                  None   
1227441                Monroe County  ...                  None   
297964                 Broome County  ...                  None   
344293                Niagara County  ...                  None   
284060            Westchester County  ...                  None   
344333   Glens Falls District Office  ...                  None   
1148111           Schenectady County  ...                  N

In [10]:
insp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1310873 entries, 0 to 1310872
Data columns (total 31 columns):
 #   Column                        Non-Null Count    Dtype  
---  ------                        --------------    -----  
 0   facility                      1310834 non-null  object 
 1   address                       1310832 non-null  object 
 2   date_of_inspection            1310834 non-null  object 
 3   violation_item                1069652 non-null  object 
 4   violation_description         1069652 non-null  object 
 5   critical_violation            1310834 non-null  object 
 6   total_critical_violations     1230072 non-null  float64
 7   total_crit_not_corrected      1229454 non-null  float64
 8   total_noncritical_violations  1230201 non-null  float64
 9   local_health_department       1310834 non-null  object 
 10  county                        1310834 non-null  object 
 11  facility_code                 1310834 non-null  object 
 12  facility_address            

- Overall there are null values in the dataset. 
-  Column `Date of inspection` datatype should be changed.

## 2. PRELIMINARY ANALYSIS

### 2.1 CHECKING DATATYPES

In [14]:
insp.sample(5)

facility                              address  \
1050175             KAISEKI SUSHI  7 WEST MERRICK ROAD,  VALLEY STREAM   
821496               TARA KITCHEN    431 LIBERTY  STREET,  SCHENECTADY   
134116                BEER MIRIAM          99 MILL HOOK  ROAD,  ACCORD   
457974          MARKETPLACE GRILL       800 JEFFERSON ROAD,  ROCHESTER   
783595   FRONTERA TACOS & TEQUILA         340 RT 211 EAST,  MIDDLETOWN   

        date_of_inspection violation_item  \
1050175         10/14/2020            15A   
821496            2/8/2012             4A   
134116           6/26/2024            14B   
457974            3/7/2024            14A   
783595           11/1/2023             8B   

                                     violation_description  \
1050175  Floors, walls, ceilings, not smooth, properly ...   
821496   Toxic chemicals are improperly labeled, stored...   
134116   Effective measures not used to control entranc...   
457974                            Insects, rodents present   
783595   In use food dispensing utensils improperly stored   

             critical_violation  total_critical_violations  \
1050175  Not Critical Violation                        0.0   
821496       Critical Violation                        3.0   
134116   Not Critical Violation                        0.0   
457974   Not Critical Violation                        0.0   
783595   Not Critical Violation                        2.0   

         total_crit_not_corrected  total_noncritical_violations  \
1050175                       0.0                           2.0   
821496                        1.0                           4.0   
134116                        0.0                           3.0   
457974                        0.0                           3.0   
783595                        0.0                           5.0   

        local_health_department  ... permitted_dba  \
1050175          Nassau  County  ...          None   
821496       Schenectady County  ...          None   
134116            Ulster County  ...          None   
457974            Monroe County  ...          None   
783595            Orange County  ...          None   

                   permitted_corp_name perm_operator_last_name  \
1050175                K & R SUSHI INC                    None   
821496   BOLLYWOOD MARSALA CORPORATION                  WAHEED   
134116                            None                    None   
457974                            None                    None   
783595            Frontera of NY, Inc.        Rodolfo Brambila   

        perm_operator_first_name nys_health_operation_id  inspection_type  \
1050175                     None                946490.0       Inspection   
821496                    ANEESA                785962.0       Inspection   
134116                      None               1118873.0       Inspection   
457974                      None                693048.0    Re-Inspection   
783595               Omar Topete               1007692.0       Inspection   

                                       inspection_comments fs_facility_state  \
1050175  Seatings : 74 Food Permit# F0885409-Expired Au...                NY   
821496                                                None                NY   
134116   ansel system inspection June 2024. Walk in coo...                NY   
457974   Reinspection to occur in 3 weeks. |Discussed w...                NY   
783595                                                None                NY   

          latitude  longitude  
1050175        NaN        NaN  
821496   42.814679 -73.939991  
134116         NaN        NaN  
457974   43.087266 -77.617177  
783595         NaN        NaN  

[5 rows x 31 columns]

In [15]:
insp.dtypes

facility                         object
address                          object
date_of_inspection               object
violation_item                   object
violation_description            object
critical_violation               object
total_critical_violations       float64
total_crit_not_corrected        float64
total_noncritical_violations    float64
local_health_department          object
county                           object
facility_code                    object
facility_address                 object
facility_city                    object
facility_postal_zipcode          object
nysdoh_gazetteer                float64
facility_municipality            object
operation_name                   object
permit_expiration_date           object
food_service_type                object
food_service_description         object
permitted_dba                    object
permitted_corp_name              object
perm_operator_last_name          object
perm_operator_first_name         object


- Everything looks good expect Columns`Date of inspection` , `Permit Expiration Date` , should be changed into datetime.

# 2.2 Duplicates

In [18]:
insp.duplicated().sum()

np.int64(350)

In [19]:
(
    insp
    .loc[insp.duplicated(keep=False)]
    .sort_values(['facility'])
)

facility                        address  \
1288367  2000 CHINESE RESTAURANT  22070 US ROUTE 11,  WATERTOWN   
484736   2000 CHINESE RESTAURANT  22070 US ROUTE 11,  WATERTOWN   
975267           4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
1301839          4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
129147           4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
...                          ...                            ...   
1259115                     None                           None   
1260173                     None                           None   
1260857                     None                           None   
1262732                     None                           None   
1268373                     None                           None   

        date_of_inspection violation_item violation_description  \
1288367          6/10/2016           None                  None   
484736           6/10/2016           None                  None   
975267           7/23/2012           None                  None   
1301839          7/23/2012           None                  None   
129147           7/23/2012           None                  None   
...                    ...            ...                   ...   
1259115               None           None                  None   
1260173               None           None                  None   
1260857               None           None                  None   
1262732               None           None                  None   
1268373               None           None                  None   

             critical_violation  total_critical_violations  \
1288367  Not Critical Violation                        0.0   
484736   Not Critical Violation                        0.0   
975267   Not Critical Violation                        0.0   
1301839  Not Critical Violation                        0.0   
129147   Not Critical Violation                        0.0   
...                         ...                        ...   
1259115                    None                        NaN   
1260173                    None                        NaN   
1260857                    None                        NaN   
1262732                    None                        NaN   
1268373                    None                        NaN   

         total_crit_not_corrected  total_noncritical_violations  \
1288367                       0.0                           0.0   
484736                        0.0                           0.0   
975267                        0.0                           0.0   
1301839                       0.0                           0.0   
129147                        0.0                           0.0   
...                           ...                           ...   
1259115                       NaN                           NaN   
1260173                       NaN                           NaN   
1260857                       NaN                           NaN   
1262732                       NaN                           NaN   
1268373                       NaN                           NaN   

           local_health_department  ... permitted_dba  \
1288367  Watertown District Office  ...          None   
484736   Watertown District Office  ...          None   
975267              Orleans County  ...          None   
1301839             Orleans County  ...          None   
129147              Orleans County  ...          None   
...                            ...  ...           ...   
1259115                       None  ...          None   
1260173                       None  ...          None   
1260857                       None  ...          None   
1262732                       None  ...          None   
1268373                       None  ...          None   

                       permitted_corp_name perm_operator_last_name  \
1288367  2000 CHINESE RESTAURANT  NY, INC.                   Jiang   
484736   2000 CHINESE RESTAURANT  NY, I

- There are `350 Duplicates` in the dataset that should be removed

## 2.3 SHAPE OF THE DATASET

In [22]:
insp.shape

(1310873, 31)

- This is a huge dataset having `1.3 million` rows or entries and `31` columns.

## 3. DETAILED ANALYSIS

#### facility (Restaurant name)

In [26]:
insp.facility.unique()

array(['WENDYS OF EG', 'STEUBEN SR DINNER CLUB - BATH',
       'MOUNT KISCO COUNTRY CLUB - POOL SNAC', ..., 'COMMISSARY-LA SALVI',
       'SHALOM UMJCA', 'STARBUCKS (NYS THRUWAY)'], dtype=object)

In [27]:
insp.facility.value_counts()

facility
DUNKIN DONUTS                          5849
Dunkin Donuts                          2285
SUBWAY                                 1993
CHINA GARDEN                           1252
DUNKIN' DONUTS                         1109
                                       ... 
KALBI HOUSE                               1
MEAT LOCKER BBQ AND BURGERS, LLC          1
CHILDTIME ON CROSSING BLVD                1
DUMPLING KING                             1
PORT WASHINGTON SD - GUGENHEIM E.S.       1
Name: count, Length: 31971, dtype: int64

#### Address

In [29]:
insp.address.value_counts()

address
3349 MONROE AVENUE,  ROCHESTER         1244
267 UNION AVENUE,  SARATOGA SPRINGS    1036
1 CROSSGATES MALL ROAD,  ALBANY        1036
9993 Alleghany ROAD,  Darien Center     989
356 Villa Roma ROAD,  Callicoon         741
                                       ... 
580 BROADWAY,  KINGSTON                   1
100 MAIN STREET,  BINGHAMTON              1
11 NEIL COURT,  OCEANSIDE                 1
33 EAST MAIN STREET,  WESTFIELD           1
175 CANADA STREET,  LAKE GEORGE           1
Name: count, Length: 31989, dtype: int64

#### Violation food item

In [31]:
insp.violation_item.value_counts()

violation_item
15A    110164
8A      98253
10B     78069
11D     77756
15B     68602
        ...  
4           1
35          1
12          1
8           1
46          1
Name: count, Length: 98, dtype: int64

#### Inspection date

In [33]:
insp.date_of_inspection.value_counts()

date_of_inspection
11/15/2023    926
11/30/2023    846
3/23/2023     831
2/1/2024      831
1/31/2024     831
             ... 
7/1/1997        1
5/7/2006        1
1/31/1999       1
1/8/2017        1
1/13/2018       1
Name: count, Length: 9463, dtype: int64

In [34]:
pd.to_datetime(insp.date_of_inspection,dayfirst = True,format='mixed')

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 06/20/0200, at position 2565

In [36]:
(
    insp.
    date_of_inspection.
    loc[2565]
)

'06/20/0200'

In [38]:
# Fixing in correct data(year) at position 2565
insp.loc[insp['date_of_inspection'] == '06/20/0200', 'date_of_inspection'] = '06/20/2000'

In [40]:
invalid_dates = insp[pd.to_datetime(insp['date_of_inspection'], errors='coerce').isnull()]['date_of_inspection']

print(invalid_dates)

30               None
2896       04/07/0005
5640       07/14/0004
46138            None
119759           None
              ...    
1287453    12/08/0002
1291687    02/20/0300
1298900    12/02/0003
1307693    09/08/0300
1310690    04/21/0300
Name: date_of_inspection, Length: 87, dtype: object


In [42]:
invalid_dates.value_counts()

date_of_inspection
04/30/0200    3
11/03/0005    2
01/20/0300    2
04/07/0005    1
11/18/0505    1
03/24/0300    1
11/13/0500    1
03/23/0006    1
04/04/0022    1
02/21/0200    1
07/14/0004    1
09/19/0003    1
11/17/0300    1
01/17/0605    1
03/19/0403    1
02/04/1003    1
10/14/0021    1
06/24/0005    1
11/23/1005    1
12/04/1002    1
01/05/1002    1
01/29/0300    1
01/10/0184    1
11/18/0300    1
03/14/0200    1
01/14/0040    1
01/23/0002    1
04/17/0003    1
04/25/0600    1
01/05/1200    1
08/04/0022    1
07/04/0022    1
10/26/1001    1
07/30/0005    1
02/24/0060    1
02/04/0020    1
05/25/0200    1
01/14/0031    1
05/16/1200    1
12/08/0002    1
02/20/0300    1
12/02/0003    1
09/08/0300    1
04/21/0300    1
Name: count, dtype: int64

#### Critical Violation 

In [44]:
insp.critical_violation.value_counts()

critical_violation
Not Critical Violation    1158252
Critical Violation         152582
Name: count, dtype: int64

#### total_critical_violations, total_crit_not_corrected, total_noncritical_violations

In [46]:
insp[['total_critical_violations', 'total_crit_not_corrected', 'total_noncritical_violations']].describe()

total_critical_violations  total_crit_not_corrected  \
count               1.230072e+06              1.229454e+06   
mean                8.127939e-01              2.876073e-02   
std                 1.436214e+00              2.947295e-01   
min                 0.000000e+00              0.000000e+00   
25%                 0.000000e+00              0.000000e+00   
50%                 0.000000e+00              0.000000e+00   
75%                 1.000000e+00              0.000000e+00   
max                 8.000000e+01              8.800000e+01   

       total_noncritical_violations  
count                  1.230201e+06  
mean                   5.250862e+00  
std                    5.160587e+00  
min                    0.000000e+00  
25%                    2.000000e+00  
50%                    4.000000e+00  
75%                    7.000000e+00  
max                    9.800000e+01

- Avearge  `violations` for restaurants is around `6.06`
- In those ,`Critical - 0.81`  , `Non-Critical-5.25`
- Average `critical cases` that are not corrected - `00.287` which is pretty low

#### local_health_department 

In [48]:
insp.local_health_department.value_counts()

local_health_department
Westchester County                                              175883
Nassau  County                                                  160850
Monroe County                                                   117206
Onondaga County                                                  79448
Glens Falls District Office                                      53669
Dutchess County                                                  52625
Rockland County                                                  47524
Orange County                                                    46475
Broome County                                                    44638
Niagara County                                                   38557
Albany County                                                    38247
Saranac Lake District Office                                     35425
Oneonta District Office                                          32182
Ulster County                                        

#### Facility muncipality

In [50]:
insp.facility_municipality.value_counts()


facility_municipality
HEMPSTEAD          76854
ROCHESTER          42360
N. HEMPSTEAD       35105
OYSTER BAY         33497
SYRACUSE           28710
                   ...  
NASSAU COUNTY          1
CATON                  1
AMBOY                  1
WETHERSFIELD           1
SCHUYLER COUNTY        1
Name: count, Length: 914, dtype: int64

#### Permit Expiry date

In [52]:
insp.permit_expiration_date.value_counts()

permit_expiration_date
12/31/2024    131697
3/31/2025     127146
1/31/2025     126668
9/30/2024     110679
12/31/2019     98571
               ...  
1/15/2020          1
8/13/2021          1
5/1/2023           1
8/17/2023          1
12/27/2024         1
Name: count, Length: 404, dtype: int64

In [54]:
pd.to_datetime(insp.permit_expiration_date,dayfirst=True,format='mixed')

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 11/30/3018, at position 271

In [56]:
invalid_permit_dates = insp[pd.to_datetime(insp['permit_expiration_date'], errors='coerce').isnull()]['permit_expiration_date']


In [58]:
invalid_permit_dates.value_counts()

permit_expiration_date
3/31/3201     30
11/30/3018    17
9/30/3024      3
12/30/2400     2
Name: count, dtype: int64

#### food service type 

In [60]:
insp.food_service_type.value_counts()

food_service_type
Food Service Establishment    1131871
Institutional Food Service     154703
SED Summer Feeding Prog.        14210
SOFA Food Service               10050
Name: count, dtype: int64

#### Inspection Type

In [62]:
insp.inspection_type.value_counts()

inspection_type
Inspection       1188410
Re-Inspection     122424
Name: count, dtype: int64

## 3.1 NULL VALUES

In [64]:
insp.isna().sum()

facility                             39
address                              41
date_of_inspection                   39
violation_item                   241221
violation_description            241221
critical_violation                   39
total_critical_violations         80801
total_crit_not_corrected          81419
total_noncritical_violations      80672
local_health_department              39
county                               39
facility_code                        39
facility_address                     45
facility_city                       172
facility_postal_zipcode             172
nysdoh_gazetteer                     39
facility_municipality                39
operation_name                       39
permit_expiration_date            15672
food_service_type                    39
food_service_description             39
permitted_dba                   1177058
permitted_corp_name              305477
perm_operator_last_name          334093
perm_operator_first_name         335787


## 4. CLEANING DATA

- Imputing null values for columns `'total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations'` using `MICE` Algorithm

In [96]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [98]:
cols =['total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations']

In [262]:

invalid_dates = [x for x in invalid_dates if x is not None]

invalid_permit_dates = [x for x in invalid_permit_dates if x is not None]
def clean_data(df,cols):
    imputer = IterativeImputer(random_state=42)
    df[cols] = imputer.fit_transform(df[cols])
    df = df[~df['permit_expiration_date'].isin(invalid_permit_dates)]
    df = df[~df['date_of_inspection'].isin(invalid_dates)]
    return(
        df
    )

In [264]:
invalid_dates

['04/07/0005',
 '07/14/0004',
 '09/19/0003',
 '04/30/0200',
 '11/18/0505',
 '03/24/0300',
 '11/13/0500',
 '03/23/0006',
 '04/04/0022',
 '02/21/0200',
 '01/17/0605',
 '11/17/0300',
 '02/04/1003',
 '03/19/0403',
 '12/04/1002',
 '11/03/0005',
 '10/14/0021',
 '06/24/0005',
 '11/23/1005',
 '01/05/1002',
 '01/20/0300',
 '04/30/0200',
 '01/29/0300',
 '01/20/0300',
 '01/10/0184',
 '11/18/0300',
 '03/14/0200',
 '01/14/0040',
 '01/23/0002',
 '04/17/0003',
 '04/30/0200',
 '04/25/0600',
 '01/05/1200',
 '08/04/0022',
 '07/04/0022',
 '10/26/1001',
 '11/03/0005',
 '07/30/0005',
 '02/24/0060',
 '02/04/0020',
 '05/25/0200',
 '01/14/0031',
 '05/16/1200',
 '12/08/0002',
 '02/20/0300',
 '12/02/0003',
 '09/08/0300',
 '04/21/0300']

In [274]:
invalid_permit_dates

['11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '9/30/3024',
 '12/30/2400',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '9/30/3024',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '9/30/3024',
 '11/30/3018',
 '12/30/2400']

In [278]:
print(len(invalid_dates),len(invalid_permit_dates))

48 52


In [268]:
clean_insp = clean_data(insp,cols)

In [270]:
clean_insp['total_noncritical_violations'].isna().sum()

np.int64(0)

In [272]:
clean_insp.shape

(1310773, 31)

- Null values in `Numerical` Columns `['total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations']` are imputed using `mice` Algorithm.
- `Incorrect Date Entries` columns are removed.

- Null values in `Categorical Columns` are imputed after checking their distribution and impute method is selcted based on their distribution or skewness.

## 5. SPLITTING THE DATA INTO TRAIN,VALIDATION,TEST DATA SET 

In [280]:
from sklearn.model_selection import train_test_split

In [284]:
X = clean_insp.drop(columns = "critical_violation")
y = clean_insp.critical_violation.copy()

In [286]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size =0.2,random_state = 42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=42)
print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(838894, 30) (838894,)
(209724, 30) (209724,)
(262155, 30) (262155,)


In [312]:
train_df = pd.DataFrame(X_train)
train_df['critical_violation'] = y_train

In [316]:
train_df.sample(5)

facility                               address  \
1074586        GREEN ACRES CAFE        133 NORTH UNION STREET,  OLEAN   
941919        KYO SUSHI & STEAK                419 ROUTE 59,  AIRMONT   
464514       CENTRE THAI BISTRO  268 ATLANTIC  AVENUE,  EAST ROCKAWAY   
234564   PUPSERIA DELI KIMBERLY          28 A URBAN AVENUE,  WESTBURY   
444421      INDIA HOUSE EXPRESS     1340 MT  HOPE  AVENUE,  ROCHESTER   

        date_of_inspection violation_item  \
1074586          2/22/2024             8A   
941919            2/6/2024            10A   
464514           7/20/2022            14A   
234564          10/11/2016            12E   
444421            4/4/2022            15A   

                                     violation_description  \
1074586  Food not protected during storage, preparation...   
941919   Food (ice) contact surfaces are improperly des...   
464514                            Insects, rodents present   
234564   Handwashing facilities inaccessible, improperl...   
444421   Floors, walls, ceilings, not smooth, properly ...   

         total_critical_violations  total_crit_not_corrected  \
1074586                        1.0                       0.0   
941919                         0.0                       0.0   
464514                         0.0                       0.0   
234564                         3.0                       0.0   
444421                         0.0                       0.0   

         total_noncritical_violations local_health_department       county  \
1074586                           3.0      Cattaraugus County  CATTARAUGUS   
941919                            4.0         Rockland County     ROCKLAND   
464514                            7.0          Nassau  County       NASSAU   
234564                            7.0          Nassau  County       NASSAU   
444421                            4.0           Monroe County       MONROE   

         ...         permitted_corp_name perm_operator_last_name  \
1074586  ...                        None                   GREEN   
941919   ...                X GLATT INC.                    SHIN   
464514   ...      CENTRE THAI BISTRO INC                    None   
234564   ...  PUPSERIA DELI KIMBERLY INC                    None   
444421   ...                        None                    None   

        perm_operator_first_name nys_health_operation_id  inspection_type  \
1074586                     PAUL                553640.0       Inspection   
941919                   SUN GUN               1093380.0    Re-Inspection   
464514                      None                688015.0       Inspection   
234564                      None                744741.0       Inspection   
444421                      None                998495.0       Inspection   

                                       inspection_comments fs_facility_state  \
1074586  Discussed proper cold holding. Food protection...                NY   
941919   Please be aware that any violation of the NYS ...                NY   
464514   Score 0/7 “B”|Permit expires Nov 30 2022|Food ...                NY   
234564   Food Manager Training Cettificates: Dora Carba...                NY   
444421                                                None                NY   

          latitude  longitude      critical_violation  
1074586  42.078451 -78.429944  Not Critical Violation  
941919   41.108317 -74.059325  Not Critical Violation  
464514   40.649345 -73.666353  Not Critical Violation  
234564   40.755035 -73.562787  Not Critical Violation  
444421         NaN        NaN  Not Critical Violation  

[5 rows x 31 columns]

In [318]:
val_df = pd.DataFrame(X_val)
val_df['critical_violation'] = y_val

In [320]:
val_df.sample(5)

facility  \
1058919             SPANKY OLIVER'S ICE CREAM   
1122062                 DUNKIN DONUTS - RT 44   
675913                SU Eggers Hall Catering   
647025     Bona-Fide-Hiawatha-Solvay OES #567   
9153     FREEPORT SD -CAROLINE G. ATKINSON ES   

                                           address date_of_inspection  \
1058919            4 STATE HIGHWAY 31 EAST,  LYONS          4/26/2013   
1122062  816-818 DUTCHESS  TURNPIKE,  POUGHKEEPSIE         10/14/2009   
675913     Eggers Hall - SU Main Campus,  Syracuse           3/3/2017   
647025                     780 Woods ROAD,  Solvay         12/12/2019   
9153                 58 W SEAMAN AVENUE,  FREEPORT          12/5/2013   

        violation_item                              violation_description  \
1058919            15B  Lighting and ventilation inadequate, fixtures ...   
1122062           None                                               None   
675913              8D  Single service items reused, improperly stored...   
647025            None                                               None   
9153                2E  Accurate thermometers not available or used to...   

         total_critical_violations  total_crit_not_corrected  \
1058919                        0.0                       0.0   
1122062                        0.0                       0.0   
675913                         0.0                       0.0   
647025                         0.0                       0.0   
9153                           1.0                       0.0   

         total_noncritical_violations local_health_department    county  ...  \
1058919                           2.0  Geneva District Office     WAYNE  ...   
1122062                           0.0         Dutchess County  DUTCHESS  ...   
675913                            3.0         Onondaga County  ONONDAGA  ...   
647025                            0.0         Onondaga County  ONONDAGA  ...   
9153                              0.0          Nassau  County    NASSAU  ...   

         permitted_corp_name perm_operator_last_name perm_operator_first_name  \
1058919                 None                 MILLER                  WILLIAM    
1122062   HIGGINS DONUTS LLC                SARDINHA                    MARIO   
675913   Syracuse University                   Damon                   Tracey   
647025   Solvay Masonic Club                Heitmann                     Mark   
9153           FREEPORT UFSD                    None                     None   

        nys_health_operation_id  inspection_type  \
1058919                326710.0       Inspection   
1122062                274916.0       Inspection   
675913                 293541.0       Inspection   
647025                 881933.0       Inspection   
9153                   573912.0       Inspection   

                                       inspection_comments fs_facility_state  \
1058919                                               None                NY   
1122062                                               None                NY   
675913                                                None                NY   
647025   Routine (2nd) inspection conducted. Establishm...                NY   
9153                                                  None                NY   

          latitude  longitude      critical_violation  
1058919        NaN        NaN  Not Critical Violation  
1122062  41.701965 -73.871772  Not Critical Violation  
675913   43.047945 -76.147448  Not Critical Violation  
647025   43.057938 -76.210595  Not Critical Violation  
9153           NaN        NaN      Critical Violation  

[5 rows x 31 columns]

In [322]:
test_df = pd.DataFrame(X_test)
test_df['critical_violation'] = y_test

In [324]:
test_df.sample(5)

facility                            address  \
1046440  MADISON COUNTY FAIRGROUNDS  1968 FAIRGROUND ROAD,  BROOKFIELD   
953676             CAMP L'MAN ACHAI       1590 PERCH LAKE ROAD,  ANDES   
1092486            TEXAS CAFE, INC.         29 BROADWAY MALL,  HORNELL   
828911       DEERFIELD COUNTRY CLUB   100 CRAIG HILL DRIVE,  BROCKPORT   
1162741           KING'S WOK BUFFET      3324  US ROUTE 11 #6,  MALONE   

        date_of_inspection violation_item  \
1046440           7/8/2016            14C   
953676           7/31/2003           None   
1092486          3/28/2024            11D   
828911            6/6/2014             8A   
1162741           1/4/2013             8A   

                                     violation_description  \
1046440  Pesticide application not supervised by a cert...   
953676                                                None   
1092486   Non food contact surfaces of equipment not clean   
828911   Food not protected during storage, preparation...   
1162741  Food not protected during storage, preparation...   

         total_critical_violations  total_crit_not_corrected  \
1046440                    0.00000                   0.00000   
953676                     0.81277                   0.02875   
1092486                    0.00000                   0.00000   
828911                     1.00000                   0.00000   
1162741                    4.00000                   0.00000   

         total_noncritical_violations       local_health_department    county  \
1046440                      6.000000                Madison County   MADISON   
953676                       5.250886       Oneonta District Office  DELAWARE   
1092486                      7.000000       Hornell District Office   STEUBEN   
828911                       3.000000                 Monroe County    MONROE   
1162741                     15.000000  Saranac Lake District Office  FRANKLIN   

         ...               permitted_corp_name perm_operator_last_name  \
1046440  ...  MADISON COUNTY FAIR ON THE TRACK                   MAYNE   
953676   ...                              None                    None   
1092486  ...                  TEXAS CAFE, INC.                  ZISSIS   
828911   ...                              None                    None   
1162741  ...                   S & B KING INC.                   JIANG   

        perm_operator_first_name nys_health_operation_id  inspection_type  \
1046440                JEFFERSON                429833.0       Inspection   
953676                      None                351588.0       Inspection   
1092486                    DIANE                343123.0       Inspection   
828911                      None                830451.0       Inspection   
1162741               YONG SHENG                527456.0       Inspection   

                                       inspection_comments fs_facility_state  \
1046440  Chlorine residual detected at a concentration ...                NY   
953676                                                None                NY   
1092486  New NYS allergen information was discussed wit...                NY   
828911                                                None                NY   
1162741                                               None                NY   

          latitude  longitude      critical_violation  
1046440  42.817361 -75.312647  Not Critical Violation  
953676   42.109461 -74.805963  Not Critical Violation  
1092486  42.326254 -77.660517  Not Critical Violation  
828911         NaN        NaN  Not Critical Violation  
1162741  44.847526 -74.319872  Not Critical Violation  

[5 rows x 31 columns]

## 5.1 Uploading datasets into POSTGRESQL

In [329]:
train_df.to_sql('training_set', engine, if_exists='append', index=False)

894

In [331]:
test_df.to_sql('test_set',engine,if_exists='append',index=False)

155

In [333]:
val_df.to_sql('val_set',engine,if_exists='append',index=False)

724